src/main/java/org/glite/authz/pep/pip/provider/X509ExtractorPIP.java

```diff
@@ -0,0 +1,312 @@
+			break;
+		    default:
+			throw new PIPProcessingException("Unknown attribute "+acceptedAttrIDs[i]+" specified");
+		}
+	    }
+	}
+
+	// Log statistics
+	log.debug("PIP parsing took "+(System.nanoTime()-t0)/1000000.0+" msec");
+
+	return pipprocessed;
+    }
+
+    /**
+     * Retrieves the end-entity certificate from a set of (subject)attributes.
+     * Note that the EMI Common XACML Profile (see
+     * https://twiki.cern.ch/twiki/bin/view/EMI/CommonXACMLProfileV1_1) sends a
+     * possibly unordered list of base64 DER-encoded certs instead of a single
+     * PEM-encoded certchain as done by the EMI WN profile
+     * (https://edms.cern.ch/document/1058175/1.0.1)
+     * @param attributes (subject) attributes to parse for EEC
+     * @return end-entity certificate
+     */
+    private X509Certificate getCertFromSubject(Set<Attribute> attributes)	{
+	// Protect against empty set
+	if (attributes==null)
+	    return null;
+
+	// Loop over all attributes, looking for ATTR_X509_ISSUER
+	for (Attribute attr: attributes) {
+	    if (attr==null)
+		continue;
+
+	    if (ATTR_KEY_INFO.equals(attr.getId()))	{
+		Set<Object> attributeValues = attr.getValues();
+		if (attributeValues==null || attributeValues.size()==0)	{
+		    log.warn("Skipping invalid key-info attr: value is empty");
+		    continue;
+		}
+
+		// Inspect the first value
+		String value = (String)attributeValues.iterator().next();
+		if (value.startsWith("-----BEGIN "))	{
+		    // assume WN profile: single PEM encoded chain. Should also
+		    // be single-valued according to https://edms.cern.ch/document/1058175/1.0.1
+		    if (attributeValues.size() > 1)
+			log.warn("key-info attr has >1 values, but is PEM, taking only first value.");
+		} else {
+		    // assume CommonXACMLProfile: base64 encoded blobs, might
+		    // not be ordered, according to
+		    // https://twiki.cern.ch/twiki/bin/view/EMI/CommonXACMLProfileV1_1
+		    // so combine into PEM string ourselves and let canl-java
+		    // loadCertificateChain() sort it for us.
+		    StringBuilder pem = new StringBuilder();
+		    for (Object attrval: attributeValues)    {
+			pem.append("-----BEGIN CERTIFICATE-----\n");
+			pem.append((String)attrval);
+			pem.append("\n-----END CERTIFICATE-----\n");
+		    }
+		    value = pem.toString();
+		}
+		// Now convert PEM-string value to a X509Certificate
+		InputStream pemReader = new ByteArrayInputStream(value.getBytes(StandardCharsets.UTF_8));
+		try {
+		    // pemReader is closed by loadCertificateChain()
+		    X509Certificate[] chain = CertificateUtils.loadCertificateChain(pemReader, Encoding.PEM);
+		    return ProxyUtils.getEndUserCertificate(chain);
+		} catch (Exception e) {
+		    // This might be a IOException, but also for invalid base64
+		    // a StringIndexOutOfBoundsException or a DecoderException
+		    log.error("Skipping invalid key-info attr: Parsing value as a certificate failed: "+e.getMessage());
```

Comment:
Do you really want that, it will give quite extensive stack traces from e.g. BouncyCastle? Why not just print the actual error message?